compose: Use tippy for stream select dropdown.

amanagr opened the pull request:
Fixes #25434
<img width="936" alt="Screenshot 2023-05-05 at 5 48 29 PM" src="https://user-images.githubusercontent.com/25124304/236456162-12a78017-cde3-40fa-ae3b-efda8c9238bd.png">

alya commented:
Thanks! A few things I'm seeing when testing in Chrome on a Mac:

1. Double scroll bars, and a rectangle on the bottom that looks like a scrollbar container (no actual scrollbar).

![Screen Shot 2023-05-05 at 10 15 07 AM](https://user-images.githubusercontent.com/2090066/236523947-44bbecbd-4764-440a-b855-4bf5df8517ab.png)
![Screen Shot 2023-05-05 at 10 16 05 AM](https://user-images.githubusercontent.com/2090066/236523965-9ed4decc-d568-4f2c-87b5-67da126c0fb2.png)

2. An extraneous rectangle around the stream selection box in dark theme.

![Screen Shot 2023-05-05 at 10 15 20 AM](https://user-images.githubusercontent.com/2090066/236524048-0139019b-de72-4399-a727-2041fdb56c12.png)

alya commented:
How did you pick the background color for the dropdown? E.g. the light theme one looks lighter to me that what we currently have. In any case, we might as well use the background colors from #25116 if we are setting them anyway?

alya commented:
Keyboard navigation of the options with up/down arrows isn't working currently, which I think is pretty critical for this one. You really want to be able to type a couple of letters, and then arrow down to the option you want.

alya commented:
Oh, and we want the search filter to be sticky, and not disappear when you scroll.

alya commented:
This is a minor point, but the alignment of the down-angle looks a little low to me.

This PR:
![Screen Shot 2023-05-05 at 10 30 26 AM](https://user-images.githubusercontent.com/2090066/236526493-d88c03e2-057e-4597-9baa-4ad24b970ccb.png)

CZO:

![Screen Shot 2023-05-05 at 10 31 11 AM](https://user-images.githubusercontent.com/2090066/236526533-8fe7f845-3c85-4638-9edb-170f44049122.png)

alya commented:
I think that's all I've got for now. :)

timabbott commented on web/src/compose_recipient.js:
```diff
@@ -3,25 +3,35 @@
 import * as util from "./util";
 
-export let compose_recipient_widget;
+export let selected_stream_name;
```
We probably want to store the selected stream by it's ID, that's robust to stream name editing automatically; would also solve the namespace problem, as the value can be either `6` or `"direct"`.

amanagr replied:
Done

alya commented:
I haven't tested live-update behavior yet (e.g., if a stream name is changed or stream privacy is updated). I don't know if this PR will affect it, but we should make sure it all works.

timabbott commented on web/src/stream_dropdown_list.js:
```diff
@@ -0,0 +1,146 @@
+import $ from "jquery";
+import {delegate} from "tippy.js";
```
I kinda don't like having a function called just `delegate` rather than `tippy.delegate`, which makes it much clearer without checking the imports what it is.

amanagr replied:
Done.

timabbott commented on web/src/server_events_dispatch.js:
```diff
@@ -527,7 +527,6 @@ export function dispatch_normal_event(event) {
                         stream_settings_ui.remove_stream(stream.stream_id);
                         if (was_subscribed) {
                             stream_list.remove_sidebar_row(stream.stream_id);
-                            compose_recipient.update_stream_dropdown_options();
```
Why is this removed?

amanagr replied:
Yeah, added back.

timabbott commented on web/src/compose_state.js:
```diff
@@ -68,14 +68,11 @@ function get_or_set(fieldname, keep_leading_whitespace, no_trim) {
 
 export function stream_name() {
-    return compose_recipient.compose_recipient_widget.value();
+    return compose_recipient.selected_stream_name;
```
If we do a stream ID value, this can just look up the stream name via `stream_data.js`.

amanagr replied:
Yeah.

timabbott commented on web/src/compose_actions.js:
```diff
@@ -221,7 +220,12 @@ export function start(msg_type, opts) {
     //
     // TODO: Move these into a conditional on message_type, using an
     // explicit "clear" function for compose_state.
```
Should this TODO be removed?

amanagr replied:
Yes, removed.

amanagr commented:
@alya addressed all your comments and since I had already implemented the new theme, you can try it out in this PR. We can easily switch back to any other colors if we need to.

Also, I additionally changed bot owner dropdown using this new library.

<img width="766" alt="Screenshot 2023-05-07 at 6 16 51 PM" src="https://user-images.githubusercontent.com/25124304/236678634-cd6b16e6-80ac-4360-811b-9938818a150e.png">

alya commented:
Thanks!

Can we update the keyboard nav so that pressing "Enter" in the search box immediately selects the top option. That's how it works now, and I think it's helpful.

alya commented:
@terpimost This PR implements #25116 specifically for the stream dropdown in the compose box. Could you please review the changes?

amanagr commented:
> Can we update the keyboard nav so that pressing "Enter" in the search box immediately selects the top option. That's how it works now, and I think it's helpful.

Done.

timabbott commented on web/src/dropdown_widget.js:
```diff
@@ -0,0 +1,134 @@
+export function setup(popover_props, get_options, item_click_callback) {
+    const on_show_callback = popover_props.on_show_callback;
+    // Since we pass `popover_props` to `tippy`, we should delete our custom properties.
+    delete popover_props.on_show_callback;
```
I'm not sure, but it might be cleaner to just have a separate section for tippy props.

amanagr replied:
Ok, makes sense.

timabbott commented on web/styles/popovers.css:
```diff
@@ -25,6 +25,20 @@
     }
 }
 
+.tippy-box[data-theme="popovers"] {
```
This name seems a bit confusing -- we already have the `light-border` theme, above, which we use for our popovers; I think we probably want to call this something more like `dropdown-widget` or `typeahead-popovers` if that's a coherent name for this theme (will typeahead look kinda similar)?

amanagr replied:
I think `dropdown-widget` sounds good. I am not sure about typeaheads.

timabbott commented on web/templates/dropdown_list.hbs:
```diff
@@ -0,0 +1,13 @@
+            {{/if}}
+        </a>
+    </li>
+{{/with}}
```
Let's start a #frontend conversation about whether we want to create a `web/templates/components` directory for this sort of thing. Can definitely be a follow-up as it's just about moving around code.

amanagr replied:
https://chat.zulip.org/#narrow/stream/6-frontend/topic/dropdown.20list.20handlebars

timabbott commented on web/src/dropdown_widget.js:
```diff
@@ -0,0 +1,134 @@
+                            // Select first item.
+                            first_item().trigger("click");
+                        }
+                        e.target.click();
```
How does this work -- it seems like we're triggering a click twice in some cases?

timabbott replied:
I was able to reproduce buggy behavior with this logic as written that is fixed with the following changes, which I've squashed in.

```
diff --git a/web/src/dropdown_widget.js b/web/src/dropdown_widget.js
index 042a3d8d02..15051d043d 100644
--- a/web/src/dropdown_widget.js
+++ b/web/src/dropdown_widget.js
@@ -51,7 +51,7 @@ export function setup(popover_props, get_options, item_click_callback) {
 
                 const $search_input = $popper.find(".dropdown-list-search-input");
                 const list_items = list_widget.get_current_list();
-                if (list_items.length === 0 && !(e.key === "Escape" || e.key === "Enter")) {
+                if (list_items.length === 0 && !(e.key === "Escape")) {
                     // Let the browser handle it.
                     return;
                 }
@@ -68,13 +68,12 @@ export function setup(popover_props, get_options, item_click_callback) {
 
                 switch (e.key) {
                     case "Enter":
-                        if (list_items.length === 0) {
-                            $(e.target).trigger("click");
-                        } else {
-                            // Select first item.
+                        if (e.target === $search_input.get(0)) {
+                            // Select first item if in search input.
                             first_item().trigger("click");
+                        } else if (list_items.length !== 0) {
+                            $(e.target).trigger("click");
                         }
-                        e.target.click();
                         e.stopPropagation();
                         e.preventDefault();
                         break;
```

amanagr replied:
Oops, thanks for fixing!

timabbott commented on web/src/dropdown_widget.js:
```diff
@@ -0,0 +1,134 @@
+            // Set focus on search input when dropdown opens.
+            setTimeout(() => {
+                $(".dropdown-list-search-input").trigger("focus");
+            });
```
Hmm, I wonder if this could be done better as an on-show callback.

amanagr replied:
I felt we want this behaviour for all the dropdowns, so it belongs here?

timabbott commented on web/src/compose_banner.ts:
```diff
@@ -137,12 +137,6 @@ export function show_stream_does_not_exist_error(stream_name: string): void {
-    $("#id_compose_select_recipient button").trigger("click");
+    // Open stream select dropdown.
+    $("#compose_select_recipient_widget").trigger("click");
 }
```
Is it still possible for this error to happen? We can deal with that as a follow-up, but now that it's not a plain-text input, I'm not sure how you get a stream that doesn't exist in there. (I guess maybe the stream get's deleted while it's open?)

amanagr replied:
> I guess maybe the stream get's deleted while it's open?

or just server and client not being in sync? It might be just better to have it than not.

timabbott commented on web/src/compose_recipient.js:
```diff
@@ -19,9 +21,17 @@ import * as stream_data from "./stream_data";
 import * as util from "./util";
 
-export let compose_recipient_widget;
+// selected_recipient_id is used store which recipient is selected:
```
"selected_recipient_id stores which recipient is selected:"

timabbott replied:
Changed this.

amanagr replied:
Thanks, my bad.

timabbott commented on web/src/compose_recipient.js:
```diff
@@ -186,36 +198,42 @@ export function on_compose_select_recipient_update(new_value) {
-        }
+        const stream_name = compose_state.stream_name();
+        stream_bar.decorate(stream_name, $stream_header_colorblock);
+        switch_message_type("stream");
```
I remember there being issues with calling this unconditionally around resetting focus in undesirable way motivated our having that `if` statement, but I've not actually reproduced any, so maybe it's fine. @evykassirer do you remember why we had that if statement to not change the message type if it was already correct?

timabbott replied:
Or maybe it's the other way around, looking at the comment for `switch_message_type("private");` -- we want to do the focus logic unconditionally and skipping it doesn't do that?

amanagr replied:
`update_placeholder_text();` is the function I need to call here regardless of `message_type`, for which `switch_message_type` seems like the right function to call.

timabbott commented on web/src/compose_recipient.js:
```diff
@@ -186,36 +198,42 @@ export function on_compose_select_recipient_update(new_value) {
         // after updating the stream.
         ui_util.place_caret_at_end($("#stream_message_recipient_topic")[0]);
-        check_stream_posting_policy_for_compose_box(new_value);
+        check_stream_posting_policy_for_compose_box(stream_name);
```
Just making a note for the potential follow-up of having a bunch of functions like this one pass around either stream IDs or `sub` objects depending on what makes sense.

timabbott commented on web/src/compose_recipient.js:
```diff
@@ -186,36 +198,42 @@ export function on_compose_select_recipient_update(new_value) {
-}
-
-export function possibly_update_dropdown_selection(old_stream_name, new_stream_name) {
+export function possibly_update_recipient_name_in_compose(old_stream_name) {
```
This function will get a loss less silly with a migration to using stream IDs here.

amanagr replied:
Yeah, done.

timabbott commented on web/src/compose_recipient.js:
```diff
@@ -244,23 +263,52 @@ function get_options_for_recipient_widget() {
+            open_compose_recipient_dropdown();
+            e.stopPropagation();
+            e.preventDefault();
+        }
```
What does this handler do? I would have thought that if you hit `Esc` to close it and then `Enter` again after, the idea would be that this would make that reopen it, but it appears the `Esc` step doesn't leave focus on this button anyway, so I'm not sure how to focus it without opening it.

amanagr replied:
Fixed.

amanagr replied:
Also, we are trying to transfer the enter press from `compose_recipient_selection_dropdown` to `compose_select_recipient_widget` here.

timabbott commented on web/src/settings_bots.js:
```diff
@@ -416,27 +415,38 @@ export function show_edit_bot_info_modal(user_id, from_user_info_popover) {
+            const $bot_owner = $("#bot_owner_dropdown_widget .bot_owner_name");
+            $bot_owner.text(user_full_name);
+            $bot_owner.attr("data-user-id", new_bot_owner_id);
+            $("#edit_bot_modal .dialog_submit_button").prop("disabled", user_full_name === null);
```
This has the bug that it gets undisabled even if you set the original value, but that was I think preexisting.

amanagr replied:
Removed current owner from the options to fix it.

timabbott commented on web/src/settings_bots.js:
```diff
@@ -416,27 +415,38 @@ export function show_edit_bot_info_modal(user_id, from_user_info_popover) {
+        }
+
         // Note: Rendering this is quite expensive in
         // organizations with 10Ks of users.
```
Is this comment still accurate? I wasn't sure whether the new code will actually do any work involving the list of options until the widget is actually opened -- if it doesn't, we may have fixed that bug?

amanagr replied:
Yeah, removed the comment. Also, `ListWidget` makes sure we don't do much work rendering it.

timabbott commented:
I fixed the one visible bug that I found. I think there's a couple UX/visual issues that we may want to adjust before releasing this:
* I don't like the background color for the "Filter" input going dark when you scroll down.
* I feel like widget could use some margins around the edges.
* https://github.com/zulip/zulip/pull/25446#pullrequestreview-1417651728 seems like we should investigate, but I don't think it impacts normal usage.

But I didn't find anything I'd describe as a material regression; I'm going to test-deploy this on chat.zulip.org and am optimistic about merging some version of this tomorrow.

timabbott commented on web/styles/dark_theme.css:
```diff
@@ -466,6 +487,18 @@
+    #compose_select_recipient_widget {
+        border-color: hsl(0deg 0% 0% / 60%);
+        background-color: hsl(0deg 0% 0% / 20%);
+    }
```
Could this be replaced by just having a `.dropdown-list-button` item just under `select` in the block 2 above here?

amanagr replied:
Yeah, but I don't like `color: inherit;` there, which is unpredictable. Although, used `.dropdown-widget-button` class here.

timabbott commented on web/styles/app_components.css:
```diff
@@ -1009,3 +1009,18 @@ div.overlay {
+        position: relative;
+        top: -1px;
+    }
+}
```
Just a small note that when rebasing to combine this with the background color PR, I needed to add this:

 +    background-color: hsl(0deg 0% 100%);

amanagr replied:
Added.

timabbott commented:
test-deploying this on chat.zulip.org.

terpimost commented:
Since it is not compose box redesign task I don't see a reason to be very picky about styles. So I just highlight the main issues.

1. With bottom alignment I don't think we should change the height of the dropdown even if the are no results in the filter. It is ok to have an empty dropdown with only filter input. In future we will show some placeholder there in case when there are no search results, like 'no such stream'. 
The current case of jumping dropdown isn't good: 
<img width="283" alt="image" src="https://user-images.githubusercontent.com/1903309/236976539-d9376a56-6db9-4390-a85f-cacc6a81c217.png">


2. icon size is too small, it should be 16px in case of svg: 
<img width="328" alt="image" src="https://user-images.githubusercontent.com/1903309/236976454-35789ee6-da1e-436a-a362-57baa1a3b0e1.png">

3. It is a bit weird to see list jumping when navigating with keyboard, at the bottom of the list I expect the next item to move the list by 1 row, instead I'm in the middle of the scrolled area: 
![CleanShot 2023-05-08 at 22 14 25](https://user-images.githubusercontent.com/1903309/236976826-897793c0-f33c-423a-b4dd-a87242be6f68.gif)

amanagr commented:
Reply to https://github.com/zulip/zulip/pull/25446#issuecomment-1539292469:
1. Done.
2. I used 13px since 16px looked too big give the font size of the streams.
before:
<img width="232" alt="Screenshot 2023-05-09 at 4 24 02 PM" src="https://github.com/zulip/zulip/assets/25124304/e2e1d36b-d51d-4fa4-b23d-8875235bdc71">
13px:
<img width="382" alt="Screenshot 2023-05-09 at 4 21 36 PM" src="https://github.com/zulip/zulip/assets/25124304/49af1efb-0975-40f9-ae59-ac2e766be793">
16px:
<img width="232" alt="Screenshot 2023-05-09 at 4 25 11 PM" src="https://github.com/zulip/zulip/assets/25124304/e3346397-f721-4f08-bc12-9177e7092e64">
3. I will open a frontend discussion about this.

amanagr commented:
@terpimost what are your thoughts on Tim's comments:


> I don't like the background color for the "Filter" input going  dark when you scroll down.
> I feel like widget could use some margins around the edges.

timabbott commented:
13px matching the left sidebar seems good for now.

timabbott commented:
Rebased to resolve merge conflicts.

timabbott commented:
I added a little commit to revert the `max-height` => `height` migration, since I agree that particular one of the suggested changes from the design threads feels like it's worse, and without it, this feels like a mergeable state to me.

I expect we'll have a half-dozen further tweaks to this component over the next week, but they should all be pretty small changes on top of it, which will be easier to review with this merged.

Thanks for all the work on this last-minute migration @amanagr!
